Refactor json

syoung-smallwisdom opened the pull request:
@kalperin 
@dephillipsmichael 

Next big massive migration. Monday I will be working on migrating BridgeApp and MotorControl so there may be a few more changes going into this.

Requires: https://github.com/Sage-Bionetworks/JsonModel-Swift/pull/1

Basically, just about everything on RSDFactory is deprecated in preference for adding protocol implementations to the base factory serializers.

TODO: Refactor RSDTask implementations, and I'm unsure at this point what is going to be the cleanest way to handle that. Will mull that over next sprint.

TODO: Finish documenting the JSON schema. Some classes are still missing this.

dephillipsmichael commented:
I appreciate the heads up on these changes coming through.  Before I review, because it is such a large code change, do you have more context for the goals of the PR in a Jira issue?

syoung-smallwisdom commented:
Yes, there is a JIRA but its very vague. Basically, we have been looking into feature parity on Android using Kotlin native. In order to use Kotlin native serialization with the default strategies for handling polymorphic types, there is are a couple of limitations that make the stuff we were doing using RSDFactory not possible. Specifically:

1. Default decoding if there isn't a "type" key
2. Mapping multiple "type" keys to the same object.
3. Combining different key values to return a mapping of type to class.

Because of this, the `RSDUIFormStep` and `RSDInputField` protocols have been deprecated and replaced with a one-to-one mapping of an `RSDStep` to a "type". Also, we moved to using a swift package called `JsonModel-Swift` to implement polymorphic implementations using the same strategy that is used by Kotlin native.

dephillipsmichael commented:
> Yes, there is a JIRA but its very vague. Basically, we have been looking into feature parity on Android using Kotlin native. In order to use Kotlin native serialization with the default strategies for handling polymorphic types, there is are a couple of limitations that make the stuff we were doing using RSDFactory not possible. Specifically:
> 
> 1. Default decoding if there isn't a "type" key
> 2. Mapping multiple "type" keys to the same object.
> 3. Combining different key values to return a mapping of type to class.
> 
> Because of this, the `RSDUIFormStep` and `RSDInputField` protocols have been deprecated and replaced with a one-to-one mapping of an `RSDStep` to a "type". Also, we moved to using a swift package called `JsonModel-Swift` to implement polymorphic implementations using the same strategy that is used by Kotlin native.

Appreciate the additional info.  I'm actually happy to hear that, I've found the nested enum types to be confusing to use and understand, and I'm often struggling on how to create them programmatically.   I won't be able to do a comprehensive review on this code, but I'll do a quick look to familiarize myself.

dephillipsmichael commented on Research/Research/AnswerTypeObjects.swift:
```diff
@@ -469,9 +532,13 @@ public struct AnswerTypeDateTime : RSDBaseAnswerType, Codable, Hashable {
     public static let defaultJsonType: JsonType = .number
-    public private(set) var type: AnswerTypeType = .measurement
+    public static let defaultType: AnswerTypeType = .measurement
+    public private(set) var objectType: AnswerTypeType = Self.defaultType
```
Capitalized "Self" is a megalomaniac :P

Jokes aside, I didn't know about that use of capitalized self, good to know.

dephillipsmichael commented on Research/Research/RSDInstructionStepObject.swift:
```diff
@@ -0,0 +1,173 @@
+import Foundation
+import JsonModel
+
+open class RSDInstructionStepObject : RSDUIStepObject, RSDInstructionStep, Encodable {
```
YESSSSS! Thank you. I was always confused that I had to make my own custom implementation of the protocol every time instead of sub-classing.

dephillipsmichael commented on Research/Research/RSDTaskResultObject.swift:
```diff
@@ -32,6 +32,7 @@
 //
 
 import Foundation
+import JsonModel
```
Just curious on the thought behind making JsonModel its own framework?

syoung-smallwisdom replied:
I wanted to be explicit about this being a reusable "framework" that extends the functionality of the Codable protocol to allow for polymorphism and documentation. It does not explicitly reference any of *our* (older) model for handling these things and can therefore be used for other projects.

dephillipsmichael commented:
I skipped over the unit tests, but gave a once over to everything else.  Seems to be correctly focused on changing the existing classes to use JsonModel properly.  Seems like it will be minimal effort to update existing apps to use it since the JSON decoding format is still the same. Nice work.
